fsm.cpp

```diff
@@ -20,50 +27,61 @@ int state3_stillness_timer = 0;
 
 void fsm::setupFSM()
 {
-    updateTimersFromFlash();
-    DEBUG_SERIAL_LOG.printf("Timers constants: countdown: %u, duration: %u, stillness: %u\r\n", COUNTDOWN_TIMER, DURATION_TIMER, STILLNESS_TIMER);
-}
-
-bool fsm::updateTimersFromFlash()
-{
-    return api.system.flash.get(COUNTDOWN_TIMER_FLASH_ADDRESS, (uint8_t *)&COUNTDOWN_TIMER, sizeof(COUNTDOWN_TIMER)) &&
+    bool success = api.system.flash.get(COUNTDOWN_TIMER_FLASH_ADDRESS, (uint8_t *)&COUNTDOWN_TIMER, sizeof(COUNTDOWN_TIMER)) &&
            api.system.flash.get(DURATION_TIMER_FLASH_ADDRESS, (uint8_t *)&DURATION_TIMER, sizeof(DURATION_TIMER)) &&
-           api.system.flash.get(STILLNESS_TIMER_FLASH_ADDRESS, (uint8_t *)&STILLNESS_TIMER, sizeof(STILLNESS_TIMER)); 
+           api.system.flash.get(STILLNESS_TIMER_FLASH_ADDRESS, (uint8_t *)&STILLNESS_TIMER, sizeof(STILLNESS_TIMER));
+    if (!success) DEBUG_SERIAL_LOG.println("Error reading flash");
+    DEBUG_SERIAL_LOG.printf("Timers constants: countdown: %u, duration: %u, stillness: %u\r\n", COUNTDOWN_TIMER, DURATION_TIMER, STILLNESS_TIMER);
 }
 
 int fsm::setCountdownTimer(unsigned int timer)
 {
-    bool success = api.system.flash.set(COUNTDOWN_TIMER_FLASH_ADDRESS, (uint8_t *)&timer, sizeof(COUNTDOWN_TIMER));
-    COUNTDOWN_TIMER = timer;
-    if (success)
-    {
-        DEBUG_SERIAL_LOG.printf("Countdown timer set to %u\r\n", COUNTDOWN_TIMER);
-        return COUNTDOWN_TIMER; 
+    if (timer <= MAX_COUNTDOWN_TIMER) {
+        bool success = api.system.flash.set(COUNTDOWN_TIMER_FLASH_ADDRESS, (uint8_t *)&timer, sizeof(COUNTDOWN_TIMER));
+        if (success)
+        {
+            COUNTDOWN_TIMER = timer;
+            DEBUG_SERIAL_LOG.printf("Countdown timer set to %u\r\n", COUNTDOWN_TIMER);
+            return COUNTDOWN_TIMER;
+        }
     }
+    DEBUG_SERIAL_LOG.printf("Failed to set countdown timer to %u\r\n", timer);
     return -1;
 }
 
 int fsm::setDurationTimer(unsigned int timer)
 {
-    bool success = api.system.flash.set(DURATION_TIMER_FLASH_ADDRESS, (uint8_t *)&timer, sizeof(DURATION_TIMER));
-    DURATION_TIMER = timer;
-    if (success)
-    {
-        DEBUG_SERIAL_LOG.printf("Duration timer set to %u\r\n", DURATION_TIMER);
-        return DURATION_TIMER;
+    if (timer >= MIN_DURATION_TIMER) {
+        bool success = api.system.flash.set(DURATION_TIMER_FLASH_ADDRESS, (uint8_t *)&timer, sizeof(DURATION_TIMER));
+        if (success)
+        {
+            DURATION_TIMER = timer;
+            DEBUG_SERIAL_LOG.printf("Duration timer set to %u\r\n", DURATION_TIMER);
+            return DURATION_TIMER;
+        }
     }
+    DEBUG_SERIAL_LOG.printf("Failed to set duration timer to %u\r\n", timer);
     return -1;
 }
 
+void fsm::resetTimers() {
+    fsm::setCountdownTimer(DEFAULT_COUNTDOWN_TIMER); 
+    fsm::setDurationTimer(DEFAULT_DURATION_TIMER);
+    fsm::setStillnessTimer(DEFAULT_STILLNESS_TIMER);
```

Comment:
I forget, in C++ is it necessary for a function to be defined before it's used? I wonder if `setStillnessTimer` should be above `resetTimers`.

Reply:
No it is not needed, as long as it's defined in the `.h` file. However, I have moved it up, as otherwise it is squished in between two related functions.